backend/server.js

```diff
@@ -125,9 +165,21 @@ sbp('okTurtles.data/set', PUBSUB_INSTANCE, createServer(hapi.listener, {
   await initDB()
   await sbp('chelonia/configure', SERVER)
   // Load the saved Chelonia state
-  const savedState = await sbp('chelonia/db/get', '_private_cheloniaState')
-  if (savedState) {
-    Object.assign(sbp('chelonia/private/state'), JSON.parse(savedState))
+  // First, get the contract index
+  const savedStateIndex = await sbp('chelonia/db/get', '_private_cheloniaState_index')
+  if (savedStateIndex) {
+    // Now, we contract the contract state by reading each contract state
+    // partition
+    const recoveredState = Object.create(null)
+    recoveredState.contracts = Object.create(null)
+    await Promise.all(savedStateIndex.split('\x00').map(async (contractID) => {
```

Comment:
It seems we're using this "index" to figure out what contractIDs we have?

Another potential approach would be to update the database API functions to return that info. For example, with FS database, you can do the equivalent of `ls head=*` and that'll return all the contractIDs. For SQlite you can do `SELECT * FROM Data WHERE key LIKE 'head=%'` (or similar), and for in-memory can probably iterate the keys.

Reply:
I thought that the model was a basic KV store, in which case listing all available keys may not be possible, or if possible, may not be desirable.

As you see in your suggestions:

* For SQLite maybe this would be OK-ish, but still you're reading *all* keys and then filtering them. I'm saying this is OK-ish because maybe internally SQLite can be told to make an index.
* For FS and memory, you have to enumerate *all* keys, process then and then filter them. Doing this is memory and resource intensive, especially if there are many keys, and even more so if the proportion of keys to get out of the total number of keys is low (which would be expected in this case).

Reply:
Ah, that's a great point. Resolving this.